h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,161 @@
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Persistence backend for GCS
+ */
+public final class PersistGcs extends Persist {
+
+  private static final Storage storage = StorageOptions.getDefaultInstance().getService();
+
+  @Override
+  public byte[] load(Value v) throws IOException {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    Log.info("Loading: " + blobId.toString());
+    return storage.get(blobId).getContent();
+  }
+
+  @Override
+  public Key uriToKey(URI uri) throws IOException {
+    final GcsBlob blob = GcsBlob.of(uri);
+    final Long contentSize = storage.get(blob.getBlobId()).getSize();
+    return GcsFileVec.make(blob.getCanonical(), contentSize);
+  }
+
+  @Override
+  public void store(Value v) throws IOException {
+    if (!v._key.home()) return;
+    final byte payload[] = v.memOrLoad();
+    final GcsBlob blob = GcsBlob.of(v._key);
+    Log.info("Storing: " + blob.toString());
+    final ByteBuffer buffer = ByteBuffer.wrap(payload);
+    storage.create(blob.getBlobInfo()).writer().write(buffer);
+  }
+
+  @Override
+  public void delete(Value v) {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    Log.info("Deleting: " + blobId.toString());
+    storage.get(blobId).delete();
+  }
+
+  @Override
+  public void cleanUp() {
+    throw H2O.unimpl();
+  }
+
+  private static final Map<String, StringCache> keyCache = new HashMap<>();
+  private static final StringCache bucketCache = new StringCache() {
+    @Override
+    List<String> update() {
+      final List<String> c = new ArrayList<>();
+      for (Bucket b : storage.list().iterateAll()) {
+        c.add(b.getName());
+      }
+      return c;
+    }
+  };
+
+  @Override
+  public List<String> calcTypeaheadMatches(String filter, int limit) {
+    final String input = GcsBlob.removePrefix(filter);
+    final String[] bk = input.split("/", 2);
+    if (bk.length == 1) {
+      return bucketCache.fetch(GcsBlob.KEY_PREFIX, bk[0], limit);
+    } else if (bk.length == 2) {
+      if (!keyCache.containsKey(bk[0])) {
+        StringCache sc = new StringCache() {
+          @Override
+          List<String> update() {
+            final List<String> cache = new ArrayList<>();
+            for (Blob b : storage.get(bk[0]).list().iterateAll()) {
+              cache.add(b.getName());
+            }
+            return cache;
+          }
+        };
+        keyCache.put(bk[0], sc);
+      }
+      return keyCache.get(bk[0]).fetch(GcsBlob.KEY_PREFIX + bk[0] + "/", bk[1], limit);
+    }
+    return new ArrayList<>(0);
+  }
+
+  @Override
+  public void importFiles(String path,
+                          String pattern,
+                          ArrayList<String> files,
+                          ArrayList<String> keys,
+                          ArrayList<String> fails,
+                          ArrayList<String> dels) {
+    final String bk[] = GcsBlob.removePrefix(path).split("/", 2);
+    final Bucket bucket = storage.get(bk[0]);
```

Comment:
If there is no pattern and path leads only to one file (which is exposed to public), however the whole bucket is private, this method fails, as it tries to scan the bucket without credentials. This is an unlikely scenario, but something I discovered.

GCS allows users to download single files from the whole bucket by having bucket identifier and the blob identifier, while the rest of bucket's content remains private. Currently, we do not support such scenario.

Reply:
Fixed in new PR https://github.com/h2oai/h2o-3/pull/2119